### ISSUE TITLE
Undo removes all entered text

### ISSUE THREAD

**Issue by raydac** [bobbylight/RSyntaxTextArea#372]
I have written a small test where just a form contains RSyntaxTextArea component. As I see undo/redo should work just out of the box but if I press ctrl+z then all entered text (which contains multiple lines) is removed like the component doesn't have any saved intermediate undo steps. May be I should turn on something in RSyntaxTextArea to save intermediate undo steps?

**Comment by bobbylight** [comment]
RSTA's undo will delete all contiguous text.  If you moved the caret to another location and typed, it would trigger a separate undo action.

Can you confirm this is the behavior you see?

I see other code editors (at least IntelliJ and VS Code) start a new undo when pressing Enter as well.  I'm guessing this is what you expected but aren't seeing?

**Comment by raydac** [comment]
I made some small screen-cast, may I be have skipped something but looks like there is no new undo for enter pressing, the test program is visible in left part

https://user-images.githubusercontent.com/3008810/110341171-ba76cc80-8032-11eb-9acf-db3febad1016.mp4

**Comment by bobbylight** [comment]
Yes, that's the current behavior - all contiguous text, including line endings, are treated as a single undo.  I'll leave this open as a feature request for a new undo to restart on Enter presses.

**Comment by raydac** [comment]
and is there any way for programmers to replace undo manager in RSyntaxTextArea by custom one? looks like that use of RUndoManager is hard-coded

**Comment by bobbylight** [comment]
You'd have to create a custom subclass of [RUndoManager](https://github.com/bobbylight/RSyntaxTextArea/blob/master/RSyntaxTextArea/src/main/java/org/fife/ui/rtextarea/RUndoManager.java) that reimplements the logic in [undoableEditHappened()](https://github.com/bobbylight/RSyntaxTextArea/blob/master/RSyntaxTextArea/src/main/java/org/fife/ui/rtextarea/RUndoManager.java#L138) so that newlines start a new compound edit.  This should be straightforward since `RUndoManager` just wraps Swing's standard `javax.swing.undo` package's functionality.

Then inject an instance of your custom `RUndoManager` subclass via a subclass of `RSyntaxTextArea` with [createUndoManager()](https://github.com/bobbylight/RSyntaxTextArea/blob/master/RSyntaxTextArea/src/main/java/org/fife/ui/rtextarea/RTextArea.java#L549) overridden.

**Comment by bobbylight** [comment]
Reopening since I meant to actually look at implementing this in the library.

**Comment by Sothatsit** [comment]
> You'd have to create a custom subclass of [RUndoManager](https://github.com/bobbylight/RSyntaxTextArea/blob/master/RSyntaxTextArea/src/main/java/org/fife/ui/rtextarea/RUndoManager.java) that reimplements the logic in [undoableEditHappened()](https://github.com/bobbylight/RSyntaxTextArea/blob/master/RSyntaxTextArea/src/main/java/org/fife/ui/rtextarea/RUndoManager.java#L138) so that newlines start a new compound edit. This should be straightforward since `RUndoManager` just wraps Swing's standard `javax.swing.undo` package's functionality.
> 
> Then inject an instance of your custom `RUndoManager` subclass via a subclass of `RSyntaxTextArea` with [createUndoManager()](https://github.com/bobbylight/RSyntaxTextArea/blob/master/RSyntaxTextArea/src/main/java/org/fife/ui/rtextarea/RTextArea.java#L549) overridden.

Unfortunately this isn't so simple due to sub-classes not being able to access `compoundEdit`. I figured out a hack using `beginInternalAtomicEdit` and `endInternalAtomicEdit` though that allows newlines to be counted as their own actions:
```java
    private final RTextArea textArea;

    public MyUndoManager(RTextArea textArea) {
        super(textArea);
        this.textArea = textArea;
    }

    @Override public void undoableEditHappened(UndoableEditEvent e) {
        UndoableEdit edit = e.getEdit();

        if ("addition".equals(edit.getPresentationName())) {
            try {
                String inserted = textArea.getText(textArea.getCaretPosition() - 1, 1);
                if ("\n".equals(inserted)) {
                    // Create an edit action just for the newline.
                    beginInternalAtomicEdit();
                    super.undoableEditHappened(e);
                    endInternalAtomicEdit();
                    return;
                }
            } catch (BadLocationException ignored) {
                // This shouldn't happen...
            }
        }
        super.undoableEditHappened(e);
    }
```
